Core 42541 - change get decisions functionality

ninaceban opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

<!--- Describe your changes in detail -->
Changes related to get and render personalization decisions:
- renamed `viewStart` to `renderDecisions`
- renamed `scopes` to `decisionScopes`
- return decisions at the event command:
- if `renderDecisions` is true we add magical param `__view__` to `decisionScopes`. In the `event::onResponse` we render `dom-actions` decisions items. We return decisions with items that are not `dom-action` in the `event::command` promise.
- if `renderDecisions` is false or not defined and `decisionScopes` contains scopes, we return decisions in the `event:command` promise.

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which does not add functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html) or I'm an Adobe employee.
- [ ] I have submitted a [documentation](https://github.com/AdobeDocs/alloy-docs) pull request or no changes are needed.
- [x] I have made any necessary test changes and all tests pass.
- [x] I have run the Sandbox successfully.

Aaronius commented:
I think you're trying to get rid of `viewStart` entirely in this PR, which is great, but still needs some work. Do a full search for `viewStart` in the repo and you'll see what still needs to be addressed. I would expect to no longer see `viewStart` anywhere in the repo. Some references that remain are just comments, but others are more important like in `validateUserEventOptions.js`.

Aaronius commented on src/components/DataCollector/index.js:
```diff
@@ -25,11 +25,11 @@ const createDataCollector = ({ eventManager, logger }) => {
             mergeId,
-            scopes = []
+            renderDecisions = false,
+            decisionsScopes = []
```
Curious to know how others feel about this name. My initial reaction is that `decisionScopes` feels better than `decisionsScopes`.

jfkhoury replied:
@Aaronius  I like `decisionScopes` because I can pronounce it ;)

Aaronius replied:
@ninaceban Mind changing it to `decisionScopes`? I think grammatically it probably makes more sense and @jfkhoury and I will be able to pronounce it. 😄

ninaceban replied:
ok, I'll change that. thank you

Aaronius commented on src/components/DataCollector/index.js:
```diff
@@ -53,13 +53,9 @@ const createDataCollector = ({ eventManager, logger }) => {
 
           const details = {
-            isViewStart: viewStart
+            renderDecisionsEnabled: renderDecisions,
```
If we prefer the `Enabled` suffix for ourselves, should we prefer the `Enabled` suffix on the public API on behalf of our customers? Should it be `decisionRenderingEnabled` instead of `renderDecisionsEnabled`? Looking at some of the other configuration options we have:

* clickCollectionEnabled
* cookieDestinationsEnabled
* debugEnabled

I think `decisionRenderingEnabled` might be the most appropriate, though one could make the argument that `debugEnabled` should be `debuggingEnabled`. I hate naming things.

ninaceban replied:
I think that options passed in the event command should be named `renderDecisions`,`viewStart` etc., anything related to configuration settings should be suffixed with `enabled` and we should follow the convention like `debuggingEnabled`, `cookieDestinationEnabled` etc.

Aaronius replied:
👍  I do like `renderDecisions` more than `decisionRenderingEnabled`. Maybe I like it more because "enabled" sounds like a long-running configuration option rather than whether something should be done a single time (in this case, decisions should be rendered). This would seem to agree with what you've stated.

I think I'm okay with the configuration options as they are though.

ninaceban replied:
ok, then I'll rename it to `renderDecisions`

Aaronius commented on test/unit/specs/components/Personalization/index.spec.js:
```diff
@@ -10,185 +10,94 @@ OF ANY KIND, either express or implied. See the License for the specific languag
-    expect(result.length).toEqual(0);
+    return expectAsync(deferred.promise).toBeResolvedTo({
+      decisions: SCOPES_FOO1_FOO2_DECISIONS
+    });
```
Nothing asynchronous is actually happening during this process, so there shouldn't be any need for a promise. I think you can re-work this to:

```
let result;
const onResponse = func => {
  result = func({ response });
};

personalization.lifecycle.onBeforeEvent({
  event,
  renderDecisionsEnabled,
  decisionsScopes,
  onResponse
});

expect(result.decisions).toEqual(expectedResponse);
```

I believe this applies to the other test in this file as well.

Aaronius commented on test/unit/specs/components/Personalization/index.spec.js:
```diff
@@ -10,185 +10,94 @@ OF ANY KIND, either express or implied. See the License for the specific languag
-    ).toThrow();
+    return expectAsync(deferred.promise).toBeResolvedTo({
+      decisions: expectedResponse
+    });
```
@ninaceban as you start to add more specs for personalization, I think you'll find that breaking down `Personalization/index.js` into smaller modules that get injected and then mocking those modules will provide a simpler way to test. See the Context or Identity components for examples. I think we'll talk about this in the engineering check-in today as well.

ninaceban replied:
yes that makes sense, thank you Aaron

ninaceban commented on src/components/Personalization/index.js:
```diff
@@ -34,32 +29,27 @@ const mergeMeta = (event, meta) => {
-    decision => decision.scope || PAGE_WIDE_SCOPE
-  );
+    if (isNonEmptyArray(decisionItems)) {
+      const newDecision = {};
```
ok, thank you

ninaceban commented:
@Aaronius could you please take a look, it is not done yet but I want to know if I'm on the right path. Thank you

Aaronius commented on src/components/Personalization/createComponent.js:
```diff
@@ -0,0 +1,84 @@
+        mergeQuery(event, createQueryDetails(scopes));
+
+        onResponse(({ response }) =>
+          onResponseHandler({ renderDecisions, response, modules, logger })
```
Notice here that `modules` and `logger` are things that are not specific to `createComponent.js`. This is a sign that `modules` and `logger` may be things you could inject into a `createOnResponseHandler` module up in `index.js`, which would create an `onResponseHandler`, which you could then inject into `createComponent.js`. The easiest way to illustrate would probably be to open a PR showing you what that might look like.

ninaceban replied:
thank you, I've got the idea :)

ninaceban commented on src/components/Personalization/createComponent.js:
```diff
@@ -0,0 +1,84 @@
+import { initDomActionsModules } from "./dom-actions";
+import collectClicks from "./dom-actions/clicks/collectClicks";
+import { hideContainers, showContainers } from "./flicker";
+import { mergeMeta, mergeQuery, createQueryDetails } from "./event";
```
thank you

ninaceban commented on src/components/Personalization/createComponent.js:
```diff
@@ -0,0 +1,84 @@
+
+import { noop } from "../../utils";
+import { hasScopes, isAuthoringModeEnabled, getDecisionScopes } from "./utils";
+import { initDomActionsModules } from "./dom-actions";
```
sure, thank you

ninaceban commented:
@Aaronius I have extracted all dependencies as separate modules and added tests for those. If you could take a look again. There are some other changes that I would do, in the next follow up PR. Thank you

Aaronius commented on src/components/Personalization/createExecuteDecisions.js:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2019 Adobe. All rights reserved.
+Copyright 2020 Adobe. All rights reserved.
```
If your editor is changing the year automatically, I don't think this is a problem, but if you're taking the time to manually change the year, you don't need to. Legal has said that the original year is fine and doesn't need to be updated as contents of the file change.

ninaceban replied:
I see, I didn't know that! thank you!

Aaronius commented on test/unit/specs/components/Personalization/createCollect.spec.js:
```diff
@@ -0,0 +1,38 @@
+    eventManager = {
+      sendEvent: jasmine.createSpy("sendEvent"),
+      createEvent: jasmine.createSpy("createEvent").and.returnValue(event)
+    };
```
If you prefer, you can do this:

```
eventManager = jasmine.createSpyObj("eventManager", {
  sendEvent: undefined,
  createEvent: event
});
```

ninaceban replied:
thank you

Aaronius commented on test/unit/specs/components/Personalization/createExecuteDecisions.spec.js:
```diff
@@ -0,0 +1,84 @@
+    );
+  });
+
+  it("shouldn't trigger executeActions", () => {
```
I recommend adding more context here, like `shouldn't trigger executeActions when provided with empty array of actions` or something like that, just so it's easier to understand the scenario without sifting through the code.

ninaceban replied:
thank you

Aaronius commented on test/unit/specs/components/Personalization/createOnClickHandler.spec.js:
```diff
@@ -0,0 +1,39 @@
+    const clickedElement = "foo";
+    handleOnClick({ event, clickedElement });
+    expect(collectClicks).toHaveBeenCalledWith(
+      jasmine.any(Function),
```
Because this function was created by the module, I recommend adding a test to make sure that the function does what you expect it to do. Specifically, it should call `mergeMeta` with `event` and `meta` and return the result.

ninaceban replied:
there is a proposal to change the clickTracking functionality, and I would rather add these changes then, if it will still apply.

Aaronius commented on test/unit/specs/components/Personalization/createOnResponseHandler.spec.js:
```diff
@@ -0,0 +1,74 @@
+
+    expect(extractDecisions).toHaveBeenCalledWith(response);
+    expect(showContainers).toHaveBeenCalled();
+    expect(executeDecisions).toHaveBeenCalled();
```
Assert that `executeDecisions` was called with `renderableDecisions`.

ninaceban replied:
thank you

Aaronius commented on test/unit/specs/components/Personalization/createOnResponseHandler.spec.js:
```diff
@@ -0,0 +1,74 @@
+    expect(result).toEqual(expectedResult);
+  });
+
+  it("should not trigger executeDecisions and should return all decisions", () => {
```
Just add a little context here, like `when renderDecisions is false`.

ninaceban replied:
ok, thank you

Aaronius commented on test/unit/specs/components/Personalization/index.spec.js:
```diff
@@ -10,96 +10,4 @@ OF ANY KIND, either express or implied. See the License for the specific languag
-    });
-  });
-});
+describe("Personalization", () => {});
```
You can delete this file entirely. The script that runs that checks to see if a spec files exists for each source file ignores `index.js` files.

ninaceban replied:
ok, thanks

Aaronius commented on test/unit/specs/components/Personalization/createComponent.spec.js:
```diff
@@ -10,94 +10,247 @@ OF ANY KIND, either express or implied. See the License for the specific languag
+      getDecisionScopes,
+      mergeQuery,
+      createQueryDetails
+    });
```
If you want, you can extract this out to a `build` function like was done here: https://github.com/adobe/alloy/blob/b7fc7d8e68eeb7cec68201c9205ef83584a638bb/test/unit/specs/components/Privacy/createComponent.spec.js#L37-L46

Completely your choice though.

ninaceban replied:
ok, thank you

Aaronius commented:
I noticed there are still a couple places where "view start" is mentioned that need to be cleaned up. One is in About.js:

![image](https://user-images.githubusercontent.com/210820/79769862-e0479580-82e9-11ea-97e7-8098a28e4af6.png)

Another is createEventManager.js:

![image](https://user-images.githubusercontent.com/210820/79769906-ef2e4800-82e9-11ea-937d-43e1f5ee9f1d.png)

ninaceban commented:
@Aaronius I have created another branch with the changes since this one had too many conflicts with the master. Changes related to `onClickHandler` I'll address in another PR when I'll work on `click tracking`  and `notification` functionality since there might be some code design changes.
Here is the new PR: https://github.com/adobe/alloy/pull/409

Aaronius commented:
@ninaceban can we close this PR at this point?

Aaronius commented:
This was superseded by a different PR.
